OSDOCS-9015:osp local disk deployment day 2 (#70208)
openshift/openshift-docs: modules/installation-osp-local-disk-deployment.adoc

// Module included in the following assemblies:
//
// * installing_openstack/deploying-openstack-with-rootVolume-etcd-on-local-disk.adoc

:_mod-docs-content-type: Procedure
[id="installation-osp-local-disk-deployment_{context}"]
= Deploying {rh-openstack} on local disk

.Prerequisites

* You have an OpenStack cloud with a working Cinder.

> * Your OpenStack cloud has at least 75 GB of available storage to accommodate 3 root volumes for the OpenShift control plane.

Reviewer: 75 is the absolute bare minimum. You probably want more, for example [the installer recommends at least 100GB storage per node](https://github.com/openshift/installer/blob/master/docs/user/openstack/customization.md#machine-pools).

Author: Hi @mandre that's why I say "at least 75 GB".

Reviewer: As long as it's clear that 25GB root volumes is not suited for production cluster, I'm fine.

<3> Adds the etcd devices with at least local storage of 10 GB to the cluster. You can specify values greater than 10 GB as long as the etcd device fits the Nova flavor. For example, if the Nova flavor has 15 GB, you can create the etcd device with 12 GB.
====

. Verify that the control plane machines are healthy by using the following steps:

.. Wait for the control plane machine set update to finish by running the following command:

+
[source,terminal]
----
$ oc wait --timeout=90m --for=condition=Progressing=false controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

.. Verify that the 3 control plane machine sets are updated by running the following command:

+
[source,terminal]
----
$ oc wait --timeout=90m --for=jsonpath='{.status.updatedReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

.. Verify that the 3 control plane machine sets are healthy by running the following command:

+
[source,terminal]
----
$ oc wait --timeout=90m --for=jsonpath='{.status.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

> .. Verify that the `ClusterOperators` are progressing in the cluster by running the following command:

Reviewer: are _not_ progressing, right?

Author: Yeah, this is a weird one because you want to make sure that they are progressing by verifying that they are not progressing.

// Module included in the following assemblies:
//
// * installing_openstack/deploying-openstack-with-rootVolume-etcd-on-local-disk.adoc

:_mod-docs-content-type: Procedure
[id="installation-osp-local-disk-deployment_{context}"]
= Deploying {rh-openstack} on local disk

.Prerequisites

> * You have an OpenStack cloud with a working Cinder, and it has at least 75 GB of available storage to accommodate 3 root volumes for the OpenShift control plane.
> 
> * The OpenStack cloud must be deployed with Nova ephemeral storage that uses a local storage backend and not RBD.

Reviewer 2: ```suggestion
* You have an OpenStack cloud with a working Cinder, and at least 75 GB of available storage to accommodate 3 root volumes for the OpenShift control plane.

* The OpenStack cloud is deployed with Nova ephemeral storage that uses a local storage backend and not `rbd`.
```

> * The OpenStack cloud must be deployed with Nova ephemeral storage that uses a local storage backend and not RBD.

Reviewer 3: no it's not a plugin. I wouldn't change this sentence.

.Procedure

[WARNING]
====

> This procedure can testing etcd on local disk and should not be used on production cluster because in certain cases complete loss of the control plane can occur. For more information, see "Overview of backup and restore operation" under "Backup and restore".

Reviewer 2: ```suggestion
This procedure can test etcd on the local disk and should not be used on a production cluster, as in certain cases a complete loss of the control plane can occur. For more information, see the _Backup and restore guide: Overview of the backup and restore operation_.
```

Author: I've never seen italics used like that. I _think_ that is only for external references and not for internal references. https://www.ibm.com/docs/en/ibm-style?topic=references-citations#:~:text=Follow%20these%20general%20guidelines%20for%20citing%20one%20publication%20(such%20as%20a%20book%2C%20article%2C%20or%20web%20page)%20in%20another%20publication%3A

====

> . Create a Nova flavor for the control plane with 10 GB of ephemeral disk by running the following command:

Reviewer 3: with at least 10GB; I wonder if we have other manuals on sizing a disk for etcd, for our enterprise customers with large clusters.

Reviewer 3: BTW the command is an example, RAM & CPU or other attributes can be changes. The point is to show the `--ephemeral` argument.

+
[source,terminal]
----

> $ openstack flavor create --ram 16384 --disk 25 --ephemeral 10 --vcpus 4 m1.xlarge

Reviewer 3: yes, that's actually a good idea, thanks for noticing this.
Also replace m1.xlarge by <flavor name>

----

> . Deploy a cluster with root volumes for the control plane; for example:

Reviewer 2: ```suggestion
. Deploy a cluster with root volumes for the control plane:
```

+
.Example YAML file
[source,yaml]
----
# ...
controlPlane:
  name: master
  platform:
    openstack:
      type: ${CONTROL_PLANE_FLAVOR}
      rootVolume:
        size: 25
        types:
        - ${CINDER_TYPE}
  replicas: 3
# ...
----

. Deploy the cluster you created by running the following command:
+
[source,terminal]
----

> $ openshift-install create cluster --dir <installation_directory> <1>

Reviewer 3: > `openshift` here, not `openstack` or `oc`?

I don't understand this comment. We deploy OpenShift here, I think it's fine.

----
+

> <1> For `<installation_directory>`, specify the location of your customized `./install-config.yaml file` that you created in step 2.

Reviewer 2: ```suggestion
<1> For `<installation_directory>`, specify the location of your customized `./install-config.yaml file` that you created previously.
```

Reviewer 3: or "that was previously created" if we follow @ShaunaDiaz's suggestion on "you".

+
[NOTE]
====

> Ensure that the cluster you deployed is healthy before proceeding to step 4.

Reviewer 3: We generally use this command in our CI but I'm sure there is other ways to do it:

```
# We need clusteroperators to finish progressing which means the cluster is not deploying or upgrading.
oc wait clusteroperators --all --for=condition=Progressing=false
```

> [NOTE]
> ====
> Ensure that the cluster you deployed is healthy before proceeding to step 4.
> ====

Reviewer 2: ```suggestion
[IMPORTANT]
====
Ensure that the cluster you deployed is healthy before proceeding to the next step.
====
```

> . Edit the `ControlPlaneMachineSet` (CPMS) to add the additional block ephemeral device that will be used by etcd by running the following command:

Reviewer 2: ```suggestion
. Edit the `ControlPlaneMachineSet` (CPMS) to add the additional block ephemeral device for etcd by running the following command:
```

+
[%collapsible]
====
[source,terminal]
----
$ oc patch ControlPlaneMachineSet/cluster -n openshift-machine-api --type json -p ' <1>
[
    {
      "op": "add",
      "path": "/spec/template/machines_v1beta1_machine_openshift_io/spec/providerSpec/value/additionalBlockDevices", <2>
      "value": [
        {
          "name": "etcd",
          "sizeGiB": 10,
          "storage": {
            "type": "Local" <3>
          }
        }
      ]
    }
  ]
'
----

> <1> Applies the JSON patch to the `ControlPlaneMachineSet` CR.

Reviewer 2: ```suggestion
<1> Applies the JSON patch to the `ControlPlaneMachineSet` custom resource (CR).
```

<2> Specifies the path where the `additionalBlockDevices` are added.

> <3> Adds the `etcd` devices with local storage of 10 GB to the cluster.

Reviewer 3: with at least 10 GB. Note that the size of it must fit from the Nova flavor.
If the flavor has 15 GB, you can create the device with 12 GB.
But if the flavor has 10 GB, the maximum for etcd device will be 10 GB.

Reviewer 4: I'd also be interested to test what happens if etcd grows beyond this size. How could we safely recover the cluster in this case? It could involve some interesting bootstrapping problems.

We should test and document this for GA, but I don't think it's a TP blocker.

Reviewer 3: I like Matt's point. We should capture it in this doc, we'll add that in our GA EPIC (when we have one...).
Maybe add a warning that the procedure to make etcd's storage grow above the ephemeral disk size isn't figured out yet and will be when the feature is GA.

====

> . Verify that the control plane machines are healthy:

Reviewer 3: To wait for CPMS to do its job, this is what we do in our CI, feel free to re-use the commands:

```bash
function wait_for_cpms_done() {
  echo 'INFO: Waiting for masters to be updated...'
  oc wait --timeout=5m --for=condition=Progressing=true controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
  if
    ! oc wait --timeout=90m --for=condition=Progressing=false controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster 1>/dev/null || \
    ! oc wait --timeout=90m --for=jsonpath='{.spec.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster 1>/dev/null || \
    ! oc wait --timeout=90m --for=jsonpath='{.status.updatedReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster 1>/dev/null || \
    ! oc wait --timeout=90m --for=jsonpath='{.status.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster 1>/dev/null || \
    ! oc wait --timeout=90m --for=jsonpath='{.status.readyReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster 1>/dev/null; then
      echo "ERROR: CPMS not scaled to 3 replicas"
      oc get controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
      oc describe controlplanemachineset.machine.openshift.io cluster --namespace openshift-machine-api
      oc get controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
      oc describe controlplanemachineset.machine.openshift.io cluster --namespace openshift-machine-api
      exit 1
  fi
```

> .. Verify the cluster is progressing by running the following command:

Reviewer 2: ```suggestion
.. Verify that the cluster is progressing by running the following command:
```

+
[source,terminal]
----
$ oc wait --timeout=90m --for=condition=Progressing=false controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

> .. Verify that 3 replica control plane machine sets are running by running the following command:

Reviewer 2: ```suggestion
.. Verify that the 3 replica control plane machine sets are running by running the following command:
```

+
[source,terminal]
----
$ oc wait --timeout=90m --for=jsonpath='{.spec.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

.. Verify that the 3 control plane machine sets are updated by running the following command:

+
[source,terminal]
----
$ oc wait --timeout=90m --for=jsonpath='{.status.updatedReplicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

.. Verify that the 3 control plane machine sets are healthy by running the following command:

+
[source,terminal]
----
$ oc wait --timeout=90m --for=jsonpath='{.status.replicas}'=3 controlplanemachineset.machine.openshift.io -n openshift-machine-api cluster
----

> .. Verify the `ClusterOperators` are progressing in the cluster by running the following command:

Reviewer 2: ```suggestion
.. Verify that the `ClusterOperators` are progressing in the cluster by running the following command:
```

+
[source,terminal]
----
$ oc wait clusteroperators --timeout=30m --all --for=condition=Progressing=false

> ----

Reviewer 3: yeah I didn't think of anything else so far and I browsed around in OCP CI and this is a common pattern to do this thing.

.. Verify that each of the 3 control plane machines has the additional block device you previously created by running the following script:

+
[source,bash]
----
$ cp_machines=$(oc get machines -n openshift-machine-api --selector='machine.openshift.io/cluster-api-machine-role=master' --no-headers -o custom-columns=NAME:.metadata.name) <1>

if [[ $(echo "${cp_machines}" | wc -l) -ne 3 ]]; then
  exit 1
fi <2>

for machine in ${cp_machines}; do
  if ! oc get machine -n openshift-machine-api "${machine}" -o jsonpath='{.spec.providerSpec.value.additionalBlockDevices}' | grep -q 'etcd'; then
	exit 1
  fi <3>
done
----

> <1> Retrieves the machines running in the cluster.
> <2> Checks that there are 3 machines running.
> <3> Checks that 3 `etcd` devices are running on the 3 machines.

Reviewer 2: ```suggestion
<1> Retrieves the machines running in the cluster.
<2> Checks whether 3 machines are running.
<3> Checks whether 3 `etcd` devices are running on the 3 machines.
```

> . Create a file named `98-var-lib-etcd.yaml` by using the following content:

Reviewer 2: ```suggestion
. Create a file named `98-var-lib-etcd.yaml` by using the following YAML file:
```

[WARNING]
====

> This procedure can testing etcd on local disk and should not be used on production cluster because in certain cases complete loss of the control plane can occur. For more information, see "Overview of backup and restore operation" under "Backup and restore".

Reviewer 2: see previous comment

====

> 

Reviewer 4: I wonder if we should document what the configuration below does. How about:

The following configuration will reformat the ephemeral disk created by Nova, mount it to /var/lib/etcd, and copy any previous etcd database to the new mountpoint before starting etcd. Nova creates a filesystem with the label 'ephemeral0'. We look for the existence of this filesystem and reformat it with the label 'local-etcd' if we find it.

It uses the following systemd units:

#### var-lib-etcd.mount

This defines the mountpoint and the device which will be mounted there. Because we are relying on filesystem labels, we specify the device as /dev/disk/by-label/local-etcd. This link is created automatically by udev. The mount unit has an implicit dependency on the device unit. We use this dependency to trigger the create-local-etcd.service unit below. We also rely on create-local-etcd.service to fail quickly if the device can't be created. This avoids the mount unit hanging waiting for the device to exist.

#### create-local-etcd.service

This runs as a dependency of the /dev/disk/by-label/local-etcd device. It formats the ephemeral0 device with an xfs filesystem and the label 'local-etcd'. It is skipped if the device already exists, which ensures we don't format a device with data on it.

Note that this unit intentionally fails if the current machine does not have an etcd ephemeral disk. This ensures that this configuration can be safely applied before adding ephemeral disks, or that the control plane machineset can be safely updated to remove the etcd ephemeral disk without first removing this configuration. This ensures that the /var/lib/etcd mount unit will not wait indefinitely for it to be created, and systemd will proceed immediately without the mount point. A side effect will be that systemd will list this unit as failed in that case. This is expected behaviour. To resolve this, remove this configuration after ensuring no control plane machine is using an etcd ephemeral disk.

#### migrate-to-local-etcd.service

This copies any existing local etcd data to the new mountpoint. It only runs if there is existing etcd data on the root device, but not in the new mount point.

#### relabel-var-lib-etcd.service

This ensures that the etcd mount point and any data migrated from the root device have the correct SELinux labels before etcd starts.

[%collapsible]
====
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 98-var-lib-etcd
spec:
  config:
    ignition:
      version: 3.4.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Mount local-etcd to /var/lib/etcd

          [Mount]

>           What=/dev/disk/by-label/local-etcd <1>

Reviewer 2: ```suggestion
          What=/dev/disk/by-label/local-etcd # <1>
```

Reviewer 2: update for all instances of callouts used in this YAML block

          ConditionPathExists=!/var/lib/etcd/member
          ConditionPathIsDirectory=/sysroot/ostree/deploy/rhcos/var/lib/etcd/member <5>

          [Service]
          Type=oneshot
          RemainAfterExit=yes

          ExecStart=/bin/bash -c "if [ -d /var/lib/etcd/member.migrate ]; then rm -rf /var/lib/etcd/member.migrate; fi" <6>

          ExecStart=/usr/bin/cp -aZ /sysroot/ostree/deploy/rhcos/var/lib/etcd/member/ /var/lib/etcd/member.migrate
          ExecStart=/usr/bin/mv /var/lib/etcd/member.migrate /var/lib/etcd/member <7>

          [Install]
          RequiredBy=var-lib-etcd.mount
        enabled: true
        name: migrate-to-local-etcd.service
      - contents: |
          [Unit]
          Description=Relabel /var/lib/etcd

          After=migrate-to-local-etcd.service
          Before=crio.service

          [Service]
          Type=oneshot
          RemainAfterExit=yes

          ExecCondition=/bin/bash -c "[ -n \"$(restorecon -nv /var/lib/etcd)\" ]" <8>

          ExecStart=/usr/sbin/restorecon -R /var/lib/etcd

          [Install]
          RequiredBy=var-lib-etcd.mount
        enabled: true
        name: relabel-var-lib-etcd.service
----

> <1> This must be mounted by device, not label, to ensure `systemd` generates the device dependency we use in this config to trigger filesystem creation.

Reviewer 2: ```suggestion
<1> Mount by using device and not label to ensure that `systemd` creates the device dependency you are using in this configuration to trigger the creation of the file system.
```

<2> Do not run if the file system `dev/disk/by-label/local-etcd` already exists.

> <3> If `/dev/disk/by-label/ephemeral0` doesn't exist, this will fail with an alert message.

Reviewer 2: ```suggestion
<3> Fails with an alert message, if `/dev/disk/by-label/ephemeral0` doesn't exist.
```

<4> Migrates existing data to local etcd by running after `/var/lib/etcd` is mounted, but before crio starts so etcd is not running yet.

> <5> Requires that etcd `/var/lib/etcd` is mounted and does not contain a member director, but the ostree does.

Reviewer 2: ```suggestion
<5> Requires that `/var/lib/etcd` is mounted and does not contain a member director, but the ostree does.
```

<6> Cleans up any previous migration state.
<7>  Copies and moves in separate steps to ensure atomic creation of a complete member directory.

> <8> Does a quick check of the mountpoint directory before doing a full recursive relabel. If restorecon in file path `/var/lib/etcd` will not relabel the directory it does not run the recursive relabel.

Reviewer 2: ```suggestion
<8> Performs a quick check of the mount point directory before performing a full recursive relabel. If restorecon in the file path `/var/lib/etcd` cannot rename the directory, the recursive rename is not performed.
```

> <8> Does a quick check of the mountpoint directory before doing a full recursive relabel. If restorecon in file path `/var/lib/etcd` will not relabel the directory it does not run the recursive relabel.

Reviewer 4: I would have expected the comments to be left in the configuration, but this is a style issue so I'll leave this entirely up to you. My intention was that they're available on the system if a user is debugging, though.

Would the above explanation text remove the need for this?

Author: Yes, it would mean that the user can copy the config without having the comments coming with them. We tend to prefer this from a usability standpoint.

====

. Create the new `MachineConfig` object by running the following command:
+
[source,terminal]
----
$ oc create -f 98-var-lib-etcd.yaml
----
+
[NOTE]
====

> It will take some time for the etcd to move onto the local disk of each control plane machine.

Reviewer 2: ```suggestion
Wait until the `etcd` is transferred to the local disk of each control plane machine.
```

> It will take some time for the etcd to move onto the local disk of each control plane machine.

Author: @mdbooth can we provide some more specifics here?

====

> . Verify that the `etcd` has been moved to the local disk of each control plane by running the following commands:

Reviewer 2: ```suggestion
. Verify that the `etcd` has been transferred to the local disk of each control plane by running the following commands:
```

+
.. Verify that the cluster is still updating by running the following command:
+
[source,terminal]
----
$ oc wait --timeout=45m --for=condition=Updating=false machineconfigpool/master
----
.. Verify that the cluster is ready by running the following command:
+
[source,terminal]
----
$ oc wait node --selector='node-role.kubernetes.io/master' --for condition=Ready --timeout=30s
----

> .. Verify that the `clusteroperators` are running in the cluster by running the following command:

Reviewer 5: ```suggestion
.. Verify that the cluster Operators are running in the cluster by entering the following command:
```

// Module included in the following assemblies:
//
// * installation/deploying-openstack-with-rootVolumes-etcd-on-local-disk.adoc

:_mod-docs-content-type: Procedure
[id="installation-osp-local-disk-deployment_{context}"]
= Deploying {rh-openstack} on local disk

.Prerequisites

* An OpenStack cloud with Cinder being functional and enough available storage (at least 75 GB) to accommodate 3 Root Volumes for the OpenShift control plane.

* The cloud must be deployed with `NovaEnableRBDBackend` set to `False`.

.Procedure

. Create a Nova flavor for the Control Plane which allows 10 GB of Ephemeral Disk by running the following command:
+
[source,terminal]
----

> $ openstack flavor create --ram 16384 --disk 25 --ephemeral 10 --vcpus 4 m1.xlarge

Reviewer 2: same suggestion for all the instances with long commands?

          [Service]
          Type=oneshot
          RemainAfterExit=yes
          ExecCondition=/usr/bin/test ! -d /var/lib/etcd/member
          ExecStart=/usr/sbin/setenforce 0
          ExecStart=/bin/rsync -ar /sysroot/ostree/deploy/rhcos/var/lib/etcd/ /var/lib/etcd/
          ExecStart=/usr/sbin/setenforce 1
          TimeoutSec=0

          [Install]
          WantedBy=multi-user.target graphical.target
        enabled: true
        name: sync-var-lib-etcd-to-etcd.service
      - contents: |
          [Unit]
          Description=Restore recursive SELinux security contexts
          DefaultDependencies=no
          After=var-lib-etcd.mount
          Before=crio.service

          [Service]
          Type=oneshot
          RemainAfterExit=yes
          ExecStart=/sbin/restorecon -R /var/lib/etcd/
          TimeoutSec=0

          [Install]
          WantedBy=multi-user.target graphical.target
        enabled: true
        name: restorecon-var-lib-etcd.service
----
====

. Once the yaml for `98-var-lib-etcd` is created, run the following command to create the new `MachineConfig`object:
+
[source,terminal]
----
$ oc create -f 98-var-lib-etcd.yaml
----
+
[NOTE]
====

> It will take some time for the etcd to move onto the local disk of each control plane machine

Reviewer 2: Also, period is missing at the end of the sentence.

----

. Create a file named `etcd-replace.yaml`, with the following content:
+
[%collapsible]
====
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 98-var-lib-etcd
spec:
  config:
    ignition:
      version: 3.2.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Mount /dev/vdb to /var/lib/etcd
          Before=local-fs.target
          Requires=systemd-mkfs@dev-vdb.service
          After=systemd-mkfs@dev-vdb.service var.mount

          [Mount]
          What=/dev/vdb
          Where=/var/lib/etcd
          Type=xfs
          Options=defaults,prjquota

          [Install]
          WantedBy=local-fs.target
        enabled: true
        name: var-lib-etcd.mount
----
====

. Apply the file change to remove the logic for creating and syncing the device by running the following command:
+
[source,terminal]
----
$ oc replace -f etcd-replace.yaml
----
+
[NOTE]
====
Removing the logic for creating and syncing the devices will take some time to complete on each control plane machine
====

> . Check that the `etcd-replace.yaml` file changes have been appplied by running the following commands:

Reviewer 2: ```suggestion
. Verify that the `etcd-replace.yaml` file changes have been appplied:
```

+

> .. To check if the cluster is still updating, run:

Reviewer 2: ```suggestion
.. Verify that the cluster is still updating by running the following command:
```

+
[source,terminal]
----
$ oc wait --timeout=45m --for=condition=Updating=false machineconfigpool/master
----

> .. To check if the cluster is still ready, run:

Reviewer 2: ```suggestion
.. Verify that the cluster is ready by running the following command:
```

+
[source,terminal]
----
$ oc wait node --selector='node-role.kubernetes.io/master' --for condition=Ready --timeout=30s
----

> .. To check if that the clusteroperators are running in the cluster, run:

Reviewer 2: ```suggestion
.. Verify  that the `clusteroperators` are running in the cluster by running the following command:
```

+
[source,terminal]
----
$ oc wait clusteroperators --timeout=30m --all --for=condition=Progressing=false
----

> . Verify that etcd is now stored on a local disk by running the following command:

Reviewer 2: ```suggestion
. Verify that the `etcd` is now stored on a local disk by running the following command:
```

+
[source,terminal]
----

> $ oc debug node/<master_node_name> -- df -T /host/var/lib/etcd

Reviewer 2: callout?

// Module included in the following assemblies:
//
// * installing_openstack/deploying-openstack-with-rootVolume-etcd-on-local-disk.adoc

:_mod-docs-content-type: Procedure
[id="installation-osp-local-disk-deployment_{context}"]
= Deploying {rh-openstack} on local disk

.Prerequisites

* You have an OpenStack cloud with a working Cinder, and it has at least 75 GB of available storage to accommodate 3 root volumes for the OpenShift control plane.

> * The OpenStack cloud is deployed with `NovaEnableRBDBackend` set to `False`.

Reviewer 3: we should not mention TripleO and just say that Nova ephemeral storage should not be configured with the RBD backend.

> * The OpenStack cloud is deployed with `NovaEnableRBDBackend` set to `False`.

Author: Is this ok then the way that it is? @EmilienM

> * The OpenStack cloud is deployed with `NovaEnableRBDBackend` set to `False`.

Reviewer 3: no, please write:

```
The OpenStack cloud is deployed with Nova ephemeral storage using a local storage backend and not RBD.
```

Or something similar.

> * The OpenStack cloud is deployed with `NovaEnableRBDBackend` set to `False`.

Author: Updated.

if [[ $(echo "${cp_machines}" | wc -l) -ne 3 ]]; then
  exit 1
fi <2>

for machine in ${cp_machines}; do
  if ! oc get machine -n openshift-machine-api "${machine}" -o jsonpath='{.spec.providerSpec.value.additionalBlockDevices}' | grep -q 'etcd'; then
	exit 1
  fi <3>
done
----
<1> Retrieves the machines running in the cluster.
<2> Checks that there are 3 machines running.
<3> Checks that 3 `etcd` devices are running on the 3 machines.

. Create a file named `98-var-lib-etcd.yaml` by using the following content:

+
[%collapsible]
====
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 98-var-lib-etcd
spec:
  config:
    ignition:
      version: 3.2.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Make File System on /dev/vdb
          DefaultDependencies=no
          BindsTo=dev-vdb.device
          After=dev-vdb.device var.mount
          Before=systemd-fsck@dev-vdb.service

          [Service]
          Type=oneshot
          RemainAfterExit=yes

>           ExecStart=/usr/sbin/mkfs.xfs -f /dev/vdb

Reviewer 4: From `man mkfs.xfs`:
>        -f     Force overwrite when an existing filesystem is detected on the device.  By default, mkfs.xfs will not write to the device if it suspects  that
>              there is a filesystem or partition table on the device already.

That doesn't seem right. Isn't that going to blow the filesystem away on second boot? @EmilienM?

Reviewer 3: 1. It only happens once
2. The file system is already formatted by Nova, hence the `-f`. Try without and mkfs will fail. That's because the disk comes with `vfat` by default: https://opendev.org/openstack/nova/src/commit/d0b459423dd81644e8d9382b6c87fabaa4f03ad4/nova/privsep/fs.py#L257

Author: Is this ok the way that it is or does something have to change?

Reviewer 4: @EmilienM It only happens once as long as it doesn't reboot before you update the configuration to remove it. This is dangerous, and will result in loss of the control plane node if it reboots at the wrong time.

Reviewer 4: In the worst case it could result in complete loss of the control plane, e.g. a datacenter power outage.

Reviewer 3: This procedure needs to happen on day 1, right after deploying a control plane, btw.
Also this is a TechPreview procedure, this is not officially supported yet and I believe what you're rightfully asking is that we rework that procedure before we make it GA.

While the current procedure is terrible, this is something deliverable as TechPreview IMO, by explaining that:
1. Moving etcd needs to happen RIGHT AFTER deploying the control plane, before any workload is running
2. The procedure is TechPreview and will be improved & secured in the future so customers will be able to move etcd safely on day 2.

Author: @mdbooth it sounds like we need a warning in two spots. Give me a second, and I'll push an update.

          TimeoutSec=0

          [Install]
          WantedBy=var-lib-containers.mount
        enabled: true
        name: systemd-mkfs@dev-vdb.service
      - contents: |
          [Unit]
          Description=Mount /dev/vdb to /var/lib/etcd
          Before=local-fs.target
          Requires=systemd-mkfs@dev-vdb.service
          After=systemd-mkfs@dev-vdb.service var.mount

          [Mount]
          What=/dev/vdb
          Where=/var/lib/etcd
          Type=xfs
          Options=defaults,prjquota

          [Install]
          WantedBy=local-fs.target
        enabled: true
        name: var-lib-etcd.mount
      - contents: |
          [Unit]
          Description=Sync etcd data if new mount is empty
          DefaultDependencies=no
          After=var-lib-etcd.mount var.mount
          Before=crio.service

          [Service]
          Type=oneshot
          RemainAfterExit=yes
          ExecCondition=/usr/bin/test ! -d /var/lib/etcd/member

>           ExecStart=/usr/sbin/setenforce 0

Reviewer 4: We must not do this. Why is this included? @EmilienM?

Reviewer 3: I don't know and I highly don't recommend doing that.
I don't know why it's here, I've asked to look at https://github.com/openshift/installer/blob/master/docs/user/openstack/etcd-ephemeral-disk.md and this doc doesn't have it.

Author: The doc you reference does have that @EmilienM https://github.com/openshift/installer/blob/master/docs/user/openstack/etcd-ephemeral-disk.md#:~:text=crio.service%0A%0A%20%20%20%20%20%20%20%20%20%20%5BService%5D-,Type,-%3Doneshot%0A%20%20%20%20%20%20%20%20%20%20RemainAfterExit%3Dyes%0A%20%20%20%20%20%20%20%20%20%20ExecCondition

Author: Can someone help me fix the docs, please? I don't write in this area so I'm very unfamiliar with this content. Apologies in advance if I need a bit more support.

Reviewer 3: > The doc you reference does have that
yes, my bad. I was grepping for "selinux" only.

TBH I stole this from https://docs.openshift.com/container-platform/4.14/scalability_and_performance/recommended-performance-scale-practices/recommended-etcd-practices.html#move-etcd-different-disk_recommended-etcd-practices
I'm happy if we improve this, but so far this is what QE and our CI have been testing...

Reviewer 4: We should unpublish all versions of that document 😬  I'll raise a security bug against it.

Author: @mdbooth and @EmilienM what needs to change in this document for improvements?

>           ExecStart=/bin/rsync -ar /sysroot/ostree/deploy/rhcos/var/lib/etcd/ /var/lib/etcd/

Reviewer 4: Why rsync?

Incidentally, it looks like `cp -aZ` would both copy and reset SELinux in a single operation, but I haven't been able to test.

>           ExecStart=/bin/rsync -ar /sysroot/ostree/deploy/rhcos/var/lib/etcd/ /var/lib/etcd/

Reviewer 4: This will fail if `/sysroot/ostree/deploy/rhcos/var/lib/etcd/` doesn't exist, which it may not when deploying a new control plane host with this configuration.

          ExecStart=/usr/sbin/setenforce 1
          TimeoutSec=0

          [Install]
          WantedBy=multi-user.target graphical.target
        enabled: true
        name: sync-var-lib-etcd-to-etcd.service
      - contents: |
          [Unit]
          Description=Restore recursive SELinux security contexts
          DefaultDependencies=no
          After=var-lib-etcd.mount
          Before=crio.service

          [Service]
          Type=oneshot
          RemainAfterExit=yes
          ExecStart=/sbin/restorecon -R /var/lib/etcd/
          TimeoutSec=0

          [Install]
          WantedBy=multi-user.target graphical.target
        enabled: true

>         name: restorecon-var-lib-etcd.service

Reviewer 4: This will run unconditionally on every boot, but it only needs to happen when we do the original content restoration. We could move this into the sync etcd service above.

Author: Can you explain more by commenting where above you'd like this included?

Reviewer 4: I think we need to do some engineering work here to recommend a working configuraiton.

.. Verify that the `clusteroperators` are running in the cluster by running the following command:
+
[source,terminal]
----
$ oc wait clusteroperators --timeout=30m --all --for=condition=Progressing=false
----

. Create a file named `etcd-replace.yaml` by using the following content:
+
[%collapsible]
====
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: master
  name: 98-var-lib-etcd
spec:
  config:
    ignition:
      version: 3.2.0
    systemd:
      units:
      - contents: |
          [Unit]
          Description=Mount /dev/vdb to /var/lib/etcd
          Before=local-fs.target
          Requires=systemd-mkfs@dev-vdb.service
          After=systemd-mkfs@dev-vdb.service var.mount

          [Mount]
          What=/dev/vdb
          Where=/var/lib/etcd
          Type=xfs
          Options=defaults,prjquota

          [Install]
          WantedBy=local-fs.target
        enabled: true
        name: var-lib-etcd.mount
----
====

. Remove the logic for creating and syncing the device by running the following command:
+
[source,terminal]
----
$ oc replace -f etcd-replace.yaml
----

> +

Reviewer 4: This whole section isn't required if we fix the previous section to be idempotent.

// Module included in the following assemblies:
//
// * installing_openstack/deploying-openstack-with-rootVolume-etcd-on-local-disk.adoc

:_mod-docs-content-type: Procedure
[id="installation-osp-local-disk-deployment_{context}"]
= Deploying {rh-openstack} on local disk

.Prerequisites

* You have an OpenStack cloud with a working Cinder, and it has at least 75 GB of available storage to accommodate 3 root volumes for the OpenShift control plane.

* The OpenStack cloud must be deployed with Nova ephemeral storage that uses a local storage backend and not RBD.

.Procedure

[WARNING]
====

> Failure to move to etcd immidiately after deploying a control plane and before any workload is running can result in the loss of the control plane node.

Reviewer 3: not just the control plane, but can also bring data corruption, etc. Just mention "the control plane", I think is fine.

Reviewer 4: The issue isn't that moving etcd is dangerous, it's that this specific procedure is bad. I'd prefer something like:

This migration procedure is suitable only for testing etcd on local disk. It is not suitable for use on production clusters as it risks complete loss of the control plane in certain cases. It will be replaced with a robust migration procedure before this feature is generally available.

----

.. Verify the `ClusterOperators` are progressing in the cluster by running the following command:

+
[source,terminal]
----
$ oc wait clusteroperators --timeout=30m --all --for=condition=Progressing=false
----

. Verify that each of the 3 control plane machines has the additional block device you previously created by running the following script:

+
[source,bash]
----
$ cp_machines=$(oc get machines -n openshift-machine-api --selector='machine.openshift.io/cluster-api-machine-role=master' --no-headers -o custom-columns=NAME:.metadata.name) <1>

if [[ $(echo "${cp_machines}" | wc -l) -ne 3 ]]; then
  exit 1
fi <2>

for machine in ${cp_machines}; do
  if ! oc get machine -n openshift-machine-api "${machine}" -o jsonpath='{.spec.providerSpec.value.additionalBlockDevices}' | grep -q 'etcd'; then
	exit 1
  fi <3>
done
----
<1> Retrieves the machines running in the cluster.
<2> Checks that there are 3 machines running.
<3> Checks that 3 `etcd` devices are running on the 3 machines.

. Create a file named `98-var-lib-etcd.yaml` by using the following content:

> [WARNING]

Author: @mdbooth does this warning help?

Reviewer 4: My worry is that users have been known to cargo-cult knowledge like this. If we give the impression that moving etcd is dangerous, even in TP, it might be hard to convince folks to do it in production later after we've fixed the bugs.

This warning suggests that any etcd move operation must be done on a non-production cluster. That isn't true in general, and it won't be true of the version we will have to write before GA. It is only true of *this* documented move operation because of the identified bugs.

Can we write the warning in such a way that it articulates clearly:
* The procedure documented here is useful for testing the performance of etcd on local disk on a non-production cluster
* The procedure documented here has known bugs which will eat your cluster in certain circumstances
* These bugs will not be present at GA, at which time this operation will be safe

Author: > Can we write the warning in such a way that it articulates clearly:
> 
> * The procedure documented here is useful for testing the performance of etcd on local disk on a non-production cluster
> * The procedure documented here has known bugs which will eat your cluster in certain circumstances
> * These bugs will not be present at GA, at which time this operation will be safe

@mdbooth I'm a little confused because you write that `This warning suggests that any etcd move operation must be done on a non-production cluster. That isn't true in genral...`; however, you also ask that the warning be changed to include `The procedure documented here is useful for testing the performance of etcd on local disk on a non-production cluster`.

With respect to your question about re-writing the warning. The short answer is we can do the first bullet point, the second and third we can not. We can't make promises about future performance or support. We could if you'd like write up bugs in the RNs or KIs if you'd like to warn our customers about the situation.

Author: @mdbooth docs freeze is 2/13 so we are on a very tight schedule at this point. This will have to go through peer-review again and we still need QE review.
